rust/lance/src/dataset/rowids.rs

```diff
@@ -0,0 +1,253 @@
+            "id",
+            DataType::Int32,
+            false,
+        )]));
+        let num_rows = 10u64;
+        let batch = RecordBatch::try_new(
+            schema.clone(),
+            vec![Arc::new(Int32Array::from_iter_values(0..num_rows as i32))],
+        )
+        .unwrap();
+
+        let reader = RecordBatchIterator::new(vec![Ok(batch.clone())], schema.clone());
+        let write_params = WriteParams {
+            enable_experimental_stable_row_ids: true,
+            ..Default::default()
+        };
+        let temp_dir = tempfile::tempdir().unwrap();
+        let tmp_path = temp_dir.path().to_str().unwrap();
+        let dataset = Dataset::write(reader, tmp_path, Some(write_params))
+            .await
+            .unwrap();
+
+        assert_eq!(dataset.manifest().next_row_id, num_rows);
+
+        let reader = RecordBatchIterator::new(vec![Ok(batch)], schema.clone());
+        let write_params = WriteParams {
+            mode: WriteMode::Overwrite,
+            ..Default::default()
+        };
+        let dataset = Dataset::write(reader, tmp_path, Some(write_params))
+            .await
+            .unwrap();
+
+        // Overwriting should NOT reset the row id counter.
+        assert_eq!(dataset.manifest().next_row_id, 2 * num_rows);
+
+        let index = get_row_id_index(&dataset).await.unwrap();
+        assert!(index.get(0).is_none());
+        assert!(index.get(num_rows).is_some());
+    }
+
+    #[tokio::test]
+    async fn test_row_ids_update() {
+        // Updated fragments get fresh row ids.
+        let schema = Arc::new(ArrowSchema::new(vec![ArrowField::new(
+            "id",
+            DataType::Int32,
+            false,
+        )]));
+        let num_rows = 5u64;
+        let batch = RecordBatch::try_new(
+            schema.clone(),
+            vec![Arc::new(Int32Array::from_iter_values(0..num_rows as i32))],
+        )
+        .unwrap();
+
+        let reader = RecordBatchIterator::new(vec![Ok(batch.clone())], schema.clone());
+        let write_params = WriteParams {
+            enable_experimental_stable_row_ids: true,
+            ..Default::default()
+        };
+        let dataset = Dataset::write(reader, "memory://", Some(write_params))
+            .await
+            .unwrap();
+
+        assert_eq!(dataset.manifest().next_row_id, num_rows);
+
+        let dataset = UpdateBuilder::new(Arc::new(dataset))
+            .update_where("id = 3")
+            .unwrap()
+            .set("id", "100")
+            .unwrap()
+            .build()
+            .unwrap()
+            .execute()
+            .await
+            .unwrap();
+
+        let index = get_row_id_index(&dataset).await.unwrap();
+        assert!(index.get(0).is_some());
+        // Old address is still there.
+        assert_eq!(index.get(3), Some(RowAddress::new_from_parts(0, 3)));
+        // New location is there.
+        assert_eq!(index.get(5), Some(RowAddress::new_from_parts(1, 0)));
+    }
+
```

Comment:
Do you need to test delete / append or is it good enough that update works?

Reply:
I'll test those two in a follow up PR.